### PR TITLE
drivers: MH-Z19: initial support

### PR DESCRIPTION
**Pull request by bergzand**
### Contribution description

The MH-Z19 is a relative cheap (15 euro) CO2 ppm sensor. This PR adds basic support for reading the CO2 measurement over the UART interface. Saul integration is provided with this PR.

As UART is asynchronous, I had to implement a bit of synchronisation and timeout support in the driver itself. This is the reason two mutexes are used and the reason the const keyword had to be dropped in the saul interface.

### Issues/PRs references

None

### CODE REVIEW

File: sys/auto_init/saul/auto_init_mhz19.c [hunk 1]
```diff
@@ -0,0 +1,71 @@
+/*
+ * Copyright (C) 2017 Koen Zandberg
```
Comment (basilfx) [inline sys/auto_init/saul/auto_init_mhz19.c]: 2018

File: drivers/include/mhz19.h [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+#endif
+
+/**
+ * @brief   Status and error return codes
+ */
+enum {
+    MHZ19_OK            =  0,       /**< everything was fine */
+    MHZ19_ERR_UART      = -1,       /**< error initializing the UART bus */
+    MHZ19_ERR_TIMEOUT   = -2,       /**< UART timeout on retrieving sensor data */
+    MHZ19_ERR_CHECKSUM  = -3,       /**< Checksum failure on received data */
```
Comment (basilfx) [inline drivers/include/mhz19.h]: small caps checksum

File: drivers/mhz19/include/mhz19_internals.h [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * @name    MH-Z19 Baud rate
+ *
+ * Fixed at 9600 by design.
+ */
+#define MHZ19_UART_BAUDRATE             9600
```
Comment (basilfx) [inline drivers/mhz19/include/mhz19_internals.h]: I think some whitespace can be removed. I dont think you have to align all defines in the file at the same level, but per block.

File: drivers/mhz19/include/mhz19_params.h [hunk 4]
```diff
@@ -0,0 +1,77 @@
+/*
```
Comment (basilfx) [inline drivers/mhz19/include/mhz19_params.h]: Adapt to recent (SAUL) param unification.

File: sys/auto_init/saul/auto_init_mhz19.c [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+{
+    for (unsigned i = 0; i < MHZ19_NUMOF; i++) {
+        LOG_DEBUG("[auto_init_saul] initializing mhz19 #%u\n", i);
+
+        if (mhz19_init(&mhz19_devs[i], &mhz19_params[i]) < 0) {
+            LOG_ERROR("[auto_init_saul] error initializing mhz19 #%u\n", i);
+            continue;
+        }
+
+        /* temperature */
```
Comment (basilfx) [inline sys/auto_init/saul/auto_init_mhz19.c]: Not really temperature :-)

File: tests/driver_mhz19/README.md [hunk 6]
```diff
@@ -0,0 +1,12 @@
+# MH-Z19 Driver Test
+
+## Introduction
+This test will test if the MH-Z19 C02 gas sensor is working.
```
Comment (basilfx) [inline tests/driver_mhz19/README.md]: Mention MH-Z19B as well. It works as well

**Comment by basilfx** [comment]
> The MH-Z19 is a relative cheap ...
Just like the Si70xx, would it make sense to rename this to MH-Z19x? I don't have a strong opinion on this.

> ... I had to implement a bit of synchronisation ..
I can only tell it works, but I don't have an expert opinion about this

It works, with a few changes (tested SAUL and driver test application using STK3600).

**Comment by basilfx** [comment]
It's a nice device. I know its an initial implementation, but could you implement an additional command (e.g. autocalibrate, or set range), so that we have a basis for sending commands? It would not be a blocking thing, but nice to have this already in place :)

**Comment by bergzand** [comment]
The mh-z19 combines reading and writing in the same command structure it shouldn't be to hard to modify it for write operations. I'd rather not implement the zero point and the span point commands. From what I've read it's rather easy to break the sensor via those commands.

For now I'd like to keep this PR focused on an initial implementation with only support for reading the sensor. This because the other commands are either not that useful or seem to require some black magic.

File: sys/auto_init/saul/auto_init_mhz19.c (hunk 1 above)
Comment (bergzand) [reply sys/auto_init/saul/auto_init_mhz19.c]: Fixed

File: drivers/include/mhz19.h (hunk 2 above)
Comment (bergzand) [reply drivers/include/mhz19.h]: Fixed

File: drivers/mhz19/include/mhz19_internals.h (hunk 3 above)
Comment (bergzand) [reply drivers/mhz19/include/mhz19_internals.h]: Removed some whitespace here and below this line

File: drivers/mhz19/include/mhz19_params.h (hunk 4 above)
Comment (bergzand) [reply drivers/mhz19/include/mhz19_params.h]: Should be adapted now

File: sys/auto_init/saul/auto_init_mhz19.c (hunk 5 above)
Comment (bergzand) [reply sys/auto_init/saul/auto_init_mhz19.c]: Not really no, removed :)

File: tests/driver_mhz19/README.md (hunk 6 above)
Comment (bergzand) [reply tests/driver_mhz19/README.md]: Added a small explanation about the MH-Z19B also working

**Comment by bergzand** [comment]
@basilfx I think I addressed all your comments, thanks for the review!

**Comment by basilfx** [comment]
My comments have been addressed, and I am able to run the test example/saul out-of-the box on a STK3600.

I don't feel 100% comfortable to ACK for code style. Maybe someone else can take a quick look?

**Comment by bergzand** [comment]
If someone else is going to review this, I'm mostly interested in whether my approach to tackle (sychronize) the asynchronous nature of the UART interface is acceptable :)

**Comment by gebart** [comment]
Regarding the synchronization between isr and driver: I have a hard time following when and what is supposed to happen at each synchronization point. The general idea with locking a mutex twice and unlocking from isr is sound, but it is a bit difficult to follow the code when there is more than one synchronization point in the same function (get_ppm function).
@bergzand do you think you might be able to refactor it a little to make it more readable. It might be enough to add some comments and/or factor out common statements into a separate inline function to make it less confusing.

**Comment by bergzand** [comment]
@gebart Thanks for the comments, I've refactored the code a bit to split the transmission and the PPM calculation, to hopefully clarify the locking around a single exchange a bit. Now with extra comments!

File: drivers/mhz19/mhz19.c [hunk 7]
```diff
@@ -81,35 +81,59 @@ int mhz19_init(mhz19_t *dev, const mhz19_params_t *params)
     return MHZ19_OK;
 }
 
-int mhz19_get_ppm(mhz19_t *dev, int16_t *ppm)
+/*
```
Comment (basilfx) [inline drivers/mhz19/mhz19.c]: I think you need `/**`.

File: drivers/mhz19/mhz19.c [hunk 8]
```diff
@@ -81,35 +81,59 @@ int mhz19_init(mhz19_t *dev, const mhz19_params_t *params)
...
+     */
     mutex_lock(&dev->sync);
 
-    /* Unlock mutex again to reset it */
+    /* Unlock mutex again to reset it to unlocked */
     mutex_unlock(&dev->sync);
-    /* Clean up timer */
+    /* Clean up timer in case it didn't fire yet */
     xtimer_remove(&timer);
+
```
Comment (basilfx) [inline drivers/mhz19/mhz19.c]: Remove blank line.

**Comment by basilfx** [comment]
@bergzand, any updates on this? :-)

**Comment by aabadie** [comment]
It would be sad to close a PR in such a good shape. @bergzand, can you provide some output of the test application ? If you could also squash, that would be perfect.

**Comment by kaspar030** [comment]
Followup in https://github.com/RIOT-OS/RIOT/pull/13126.

**Comment by basilfx** [comment]
#13126 is now merged.

**Pull request closed without merging**